tirocks-sys/build.rs

```diff
@@ -0,0 +1,261 @@
+    env::set_var("CMAKE_PREFIX_PATH", prefix_path);
+}
+
+fn configure_common_rocksdb_args(cfg: &mut Config, name: &str) {
+    let out_dir = format!("{}/{}", env::var("OUT_DIR").unwrap(), name);
+    std::fs::create_dir_all(&out_dir).unwrap();
+    cfg.out_dir(out_dir);
+    if cfg!(feature = "portable") {
+        cfg.define("PORTABLE", "ON");
+    }
+    if cfg!(feature = "sse") {
+        cfg.define("FORCE_SSE42", "ON");
+    }
+    cfg.define("WITH_GFLAGS", "OFF")
+        .register_dep("Z")
+        .define("WITH_ZLIB", "ON")
+        .register_dep("BZIP2")
+        .define("WITH_BZ2", "ON")
+        .register_dep("LZ4")
+        .define("WITH_LZ4", "ON")
+        .register_dep("ZSTD")
+        .define("WITH_ZSTD", "ON")
+        .register_dep("SNAPPY")
+        .define("WITH_SNAPPY", "ON")
+        .configure_arg("-Wno-dev");
+}
+
+fn figure_link_lib(dst: &Path, name: &str) {
+    if cfg!(target_os = "windows") {
+        let profile = match &*env::var("PROFILE").unwrap_or_else(|_| "debug".to_owned()) {
+            "bench" | "release" => "Release",
+            _ => "Debug",
+        };
+        println!(
+            "cargo:rustc-link-search=native={}/build/{}",
+            dst.display(),
+            profile
+        );
+    } else {
+        println!("cargo:rustc-link-search=native={}/build", dst.display());
+    }
+    println!("cargo:rustc-link-lib=static={}", name);
+}
+
+fn build_titan(build: &mut Build) {
+    let cur_dir = std::env::current_dir().unwrap();
+    let mut cfg = cmake::Config::new("titan");
+    configure_common_rocksdb_args(&mut cfg, "titan");
+    let dst = cfg
+        .define("ROCKSDB_DIR", cur_dir.join("rocksdb"))
+        .define("WITH_TITAN_TESTS", "OFF")
+        .define("WITH_TITAN_TOOLS", "OFF")
+        .build_target("titan")
+        .very_verbose(true)
+        .build();
+    figure_link_lib(&dst, "titan");
+    build.include(cur_dir.join("titan").join("include"));
+    build.include(cur_dir.join("titan"));
+}
+
+fn build_rocksdb(build: &mut Build) {
+    let target = env::var("TARGET").expect("TARGET was not set");
+    let mut cfg = Config::new("rocksdb");
+    cfg.out_dir(format!("{}/rocksdb", env::var("OUT_DIR").unwrap()));
+    if cfg!(feature = "encryption") {
+        cfg.register_dep("OPENSSL").define("WITH_OPENSSL", "ON");
+        println!("cargo:rustc-link-lib=static=crypto");
+    }
+    if cfg!(feature = "jemalloc") && NO_JEMALLOC_TARGETS.iter().all(|i| !target.contains(i)) {
+        cfg.register_dep("JEMALLOC").define("WITH_JEMALLOC", "ON");
+        println!("cargo:rustc-link-lib=static=jemalloc");
+    }
+    configure_common_rocksdb_args(&mut cfg, "rocksdb");
+    let dst = cfg
+        .define("WITH_TESTS", "OFF")
+        .define("WITH_TOOLS", "OFF")
+        .build_target("rocksdb")
+        .very_verbose(true)
+        .build();
+    figure_link_lib(&dst, "rocksdb");
+
+    if cfg!(target_os = "windows") {
+        build.define("OS_WIN", None);
```

Comment:
Have a question: as the rocksdb has already been built above, do these defines make any sense?

Reply:
Yes, it will be used in the headers included in c.h.